pkg/nsx/services/subnetbinding/tree.go

```diff
@@ -0,0 +1,221 @@
+package subnetbinding
+
+import (
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/data"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	nsxutil "github.com/vmware-tanzu/nsx-operator/pkg/nsx/util"
+)
+
+var leafType = "SubnetConnectionBindingMap"
+
+type hNode struct {
+	resourceType string
+	resourceID   string
+	bindingMap   *model.SubnetConnectionBindingMap
+	childNodes   []*hNode
+}
+
+func (n *hNode) mergeChildNode(node *hNode) {
+	if node.resourceType == leafType {
+		n.childNodes = append(n.childNodes, node)
+		return
+	}
+
+	for _, cn := range n.childNodes {
+		if cn.resourceType == node.resourceType && cn.resourceID == node.resourceID {
```

Comment:
This is linear search. In a single update/delete, is it possible that at some layer of the tree, the children nodes could be many? If each node at each layer would only have a few children, I think linear search should be OK.

For example, if we want to delete all subnetbindings, the orgs, projects, vpcs should be just a few (hundreds), but how many subnets and subnetbindings would b affected? Would linear search be an issue? Maybe we can generate some fake nodes and check how much CPU time it requires to merge all children into a big tree. I estimate that it won't be an issue because the hierarchical API at most only supports 5K children (as far as I know).

Reply:
@edwardbadboy Thanks for catching this potential performance issue. I made some tests locally: I constructed 10W SubnetConnectionBindingMaps (1 org * 10 project * 20 VPC * 100 subnet * 5 binding), the total time cost by this function is about 5s.  I think this is acceptable for now. So I would prefer to improve it in another change separately.